### ISSUE TITLE
com.hazelcast.jet.sql.impl.validate.SinkTypeCoercionTest.test_insertSelect

### ISSUE THREAD

**Issue by olukas** [hazelcast/hazelcast#18944]
_master_ (commit 12a118d8f8673407fea653834be538abdeeb1d4f)

Test `com.hazelcast.jet.sql.impl.validate.SinkTypeCoercionTest.test_insertSelect` failed with following options:
- `[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]`
- `[TestParams{id=2013, srcType=TIME, targetType=TIMESTAMP WITH TIME ZONE, value=01:42:00, targetValue=2021-06-20T01:42Z, expectedFailureRegex=null}]`

Failed on Sonar build (Oracle JDK 11):
- http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar/903/testReport/com.hazelcast.jet.sql.impl.validate/SinkTypeCoercionTest/test_insertSelect_TestParams_id_2012__srcType_TIME__targetType_TIMESTAMP__value_01_42_00__targetValue_2021_06_20T01_42__expectedFailureRegex_null__/
- http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar/903/testReport/com.hazelcast.jet.sql.impl.validate/SinkTypeCoercionTest/test_insertSelect_TestParams_id_2013__srcType_TIME__targetType_TIMESTAMP_WITH_TIME_ZONE__value_01_42_00__targetValue_2021_06_20T01_42Z__expectedFailureRegex_null__/

Stacktrace (for `[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]` as an example):
```
java.lang.AssertionError: expected:<2021-06-20T01:42> but was:<2021-06-21T01:42>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at com.hazelcast.jet.sql.impl.validate.SinkTypeCoercionTest.test_insertSelect(SinkTypeCoercionTest.java:615)
	at jdk.internal.reflect.GeneratedMethodAccessor322.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Standard output (for `[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]` as an example):
```
Started Running Test: test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]
CREATE MAPPING src TYPE TestBatch OPTIONS ('names'='v', 'types'='TIME', 'values'='01:42:00')
00:02:17,522  INFO |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [SinkTypeCoercionTest] test_insertSelect - CREATE MAPPING target TYPE IMap OPTIONS('keyFormat'='int', 'valueFormat'='java', 'valueJavaClass'='com.hazelcast.sql.support.expressions.ExpressionValue$LocalDateTimeVal')
00:02:17,523  INFO |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [SinkTypeCoercionTest] test_insertSelect - SINK INTO target SELECT 0, v, 0 FROM src
00:02:17,524 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [JetSqlBackend] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Before logical opt:
LogicalTableModify(table=[[hazelcast, public, target[projects=[0, 1, 2, 3]]]], operation=[INSERT], flattened=[false])
  LogicalProject(__key=[0], field1=[CAST($0):TIMESTAMP], key=[0], this=[null:OBJECT])
    LogicalTableScan(table=[[hazelcast, public, src[projects=[0]]]])

00:02:17,525 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [JetSqlBackend] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] After logical opt:
InsertLogicalRel(table=[[hazelcast, public, target[projects=[0, 1, 2, 3]]]], operation=[INSERT], flattened=[false])
  ProjectLogicalRel(__key=[0], field1=[CAST($0):TIMESTAMP], key=[0], this=[null:OBJECT])
    FullScanLogicalRel(table=[[hazelcast, public, src[projects=[0]]]])

00:02:17,525 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [JetSqlBackend] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] After physical opt:
InsertPhysicalRel(table=[[hazelcast, public, target[projects=[0, 1, 2, 3]]]], operation=[INSERT], flattened=[false])
  ProjectPhysicalRel(__key=[0], field1=[CAST($0):TIMESTAMP], key=[0], this=[null:OBJECT])
    FullScanPhysicalRel(table=[[hazelcast, public, src[projects=[0]]]])

00:02:17,526 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [LightMasterContext] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Start executing light job 065f-36da-c927-0001, execution graph in DOT format:
digraph DAG {
	"TestBatch[public.src]" [localParallelism=1];
	"Project" [localParallelism=default];
	"Project(IMap[public.target])" [localParallelism=default];
	"IMap[public.target]" [localParallelism=1];
	"TestBatch[public.src]" -> "Project" [queueSize=1024];
	"Project" -> "Project(IMap[public.target])" [queueSize=1024];
	"Project(IMap[public.target])" -> "IMap[public.target]" [queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
00:02:17,526 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [LightMasterContext] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Building execution plan for 065f-36da-c927-0001
00:02:17,527 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [LightMasterContext] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Built execution plans for 065f-36da-c927-0001
00:02:17,527 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [InitExecutionOperation] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Initializing execution plan for job 065f-36da-c927-0001, execution 065f-36da-c927-0001 from [127.0.0.1]:5701
00:02:17,528 DEBUG |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [JobExecutionService] test_insertSelect - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution plan for light job ID=065f-36da-c927-0001, jobName='065f-36da-c927-0001', executionId=065f-36da-c927-0001 initialized, will start the execution
00:02:17,529 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-9 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Completed execution of job '065f-36da-c927-0001', execution 065f-36da-c927-0001
00:02:17,529 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-9 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution of job '065f-36da-c927-0001', execution 065f-36da-c927-0001 completed
00:02:17,670  INFO |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [SqlTestSupport] Time-limited test - Removing 1 cached plans in SqlTestSupport.@After
00:02:17,670  INFO |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [SimpleTestInClusterSupport] Time-limited test - Ditching 0 jobs in SimpleTestInClusterSupport.@After: []
00:02:17,670  INFO |test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}]| - [SimpleTestInClusterSupport] Time-limited test - Destroying 2 distributed objects in SimpleTestInClusterSupport.@After: [hz:impl:replicatedMapService/__sql.catalog, hz:impl:mapService/target]
BuildInfo right after test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}](com.hazelcast.jet.sql.impl.validate.SinkTypeCoercionTest): BuildInfo{version='5.0-SNAPSHOT', build='20210620', buildNumber=20210620, revision=12a118d, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='5.0-SNAPSHOT', build='20210620', revision='12a118d'}}
00:02:17,670 ERROR || - [EventServiceImpl] hz.tender_leavitt.event-1406 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] hz.tender_leavitt.event-1406 caught an exception while processing:com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher@4b3ac0e8
java.lang.UnsupportedOperationException: SQL catalog must never be cleared - MapEvent{entryEventType=CLEAR_ALL, member=Member [127.0.0.1]:5701 - 0b133043-b5f9-4458-96ce-d6d3f85c1471 this, name='__sql.catalog', numberOfEntriesAffected=2}
	at com.hazelcast.jet.sql.impl.schema.MappingStorage$EntryListenerAdapter.mapCleared(MappingStorage.java:159) ~[classes/:?]
	at com.hazelcast.replicatedmap.impl.ReplicatedMapEventPublishingService.dispatchEvent(ReplicatedMapEventPublishingService.java:132) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.replicatedmap.impl.ReplicatedMapService.dispatchEvent(ReplicatedMapService.java:232) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher.run(LocalEventDispatcher.java:64) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.internal.util.executor.StripedExecutor$Worker.process(StripedExecutor.java:245) [hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.internal.util.executor.StripedExecutor$Worker.run(StripedExecutor.java:228) [hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
Hiccups measured while running test 'test_insertSelect[TestParams{id=2012, srcType=TIME, targetType=TIMESTAMP, value=01:42:00, targetValue=2021-06-20T01:42, expectedFailureRegex=null}](com.hazelcast.jet.sql.impl.validate.SinkTypeCoercionTest):'
00:02:15, accumulated pauses: 153 ms, max pause: 131 ms, pauses over 1000 ms: 0
```

**Comment by viliam-durina** [comment]
The test ran close to midnight. In SQL, when casting `TIME` to any other temporal type, the current date is added. The engine gets the current date at some point in time, and the test creates the expected value at some other time. When these two happen at a different day, the test fails...

To fix we need to assert that the actual value is this or the next day, an ugly generalization...

**Comment by viliam-durina** [comment]
Closed as duplicate of #18962. The `SinkTypeCoercionTest` was renamed to `RowAssignmentTypeCoercionTest`
